### ISSUE TITLE
getChildren() and getChildrenPage crashing for empty folders

### ISSUE THREAD

**Issue by donpwilson** [CloudRail/cloudrail-si-android-sdk#65]
I'm using 2.16.3.
I had a crash using getChildren() for both DropBox and OneDrive (others not checked. I upgraded my code to use getChildrenPage() but am still getting the crash. It used to run OK. I think the problem lies when ready an empty folder, but can't be sure. I attach the code I am using to search and example crash logs.

```javaScript  private void searchFolders(final CloudStorage service, final String folder, final String text) throws Throwable {
        if ((!isCancelled()) && (folder != null)) {
            //TODO temporary value for testing purposes
            int limit = 3;
            int offset = 0;
            boolean more = true;
            ArrayList<CloudMetaData> files = new ArrayList<>();
            while (more && (!isCancelled())) {
                List<CloudMetaData> chunk_files = service.getChildrenPage(folder, offset, limit);
                files.addAll(chunk_files);
                more = (chunk_files.size() >= limit);
                offset += limit;
                My_Utils.logW(this.getClass(), "searchFolders() folder %s offset %s limit %s more %s",
                        folder, offset, limit,more);
            }
            for (CloudMetaData file : files) {
                if (file.getFolder()) {
                    searchFolders(service, file.getPath(), text);
                } else {
                    String name = file.getName();
                    if ((name != null) && (My_Utils.getFileExtension(name).equalsIgnoreCase("jpg"))) {
                        photo_data.add(file);
                    }
                }
            }
        }
    }
```

```javaScript03-29 15:10:37.611 7128-7437/com.dpw.photoplayer I/Job_LoadCloudRailData: job-queue-worker-7b0 shouldReRunOnThrowable() run count 2 drive OneDrive
                                                                          java.lang.RuntimeException: ServiceCode Error in function processRawMeta at 23
                                                                              at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:111)
                                                                              at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45)
                                                                              at com.cloudrail.si.services.OneDrive.getChildrenPage(OneDrive.java:1244)
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:263)
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111)
                                                                              at com.birbit.android.jobqueue.Job.safeRun(Job.java:229)
                                                                              at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354)
                                                                              at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389)
                                                                              at java.lang.Thread.run(Thread.java:818)
                                                                           Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Class java.lang.Object.getClass()' on a null object reference
                                                                              at com.cloudrail.si.servicecode.commands.Push.execute(Push.java:71)
                                                                              at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:87)
                                                                              at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45) 
                                                                              at com.cloudrail.si.services.OneDrive.getChildrenPage(OneDrive.java:1244) 
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:263) 
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111) 
                                                                              at com.birbit.android.jobqueue.Job.safeRun(Job.java:229) 
                                                                              at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354) 
                                                                              at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389) 
                                                                              at java.lang.Thread.run(Thread.java:818) 
```
```javaScript03-29 15:10:39.941 7128-7437/com.dpw.photoplayer I/Job_LoadCloudRailData: job-queue-worker-7b0 shouldReRunOnThrowable() run count 1 drive Dropbox
                                                                          java.lang.RuntimeException: ServiceCode Error in function getChildrenPage at 25
                                                                              at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:111)
                                                                              at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45)
                                                                              at com.cloudrail.si.services.Dropbox.getChildrenPage(Dropbox.java:1029)
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:263)
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:274)
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111)
                                                                              at com.birbit.android.jobqueue.Job.safeRun(Job.java:229)
                                                                              at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354)
                                                                              at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36)
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389)
                                                                              at java.lang.Thread.run(Thread.java:818)
                                                                           Caused by: java.lang.ArrayIndexOutOfBoundsException: length=12; index=-11
                                                                              at java.util.ArrayList.get(ArrayList.java:310)
                                                                              at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:200)
                                                                              at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:240)
                                                                              at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:240)
                                                                              at com.cloudrail.si.servicecode.Sandbox.getVariable(Sandbox.java:282)
                                                                              at com.cloudrail.si.servicecode.Sandbox.getVariable(Sandbox.java:259)
                                                                              at com.cloudrail.si.servicecode.commands.Get.execute(Get.java:44)
                                                                              at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:87)
                                                                              at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45) 
                                                                              at com.cloudrail.si.services.Dropbox.getChildrenPage(Dropbox.java:1029) 
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:263) 
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:274) 
                                                                              at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111) 
                                                                              at com.birbit.android.jobqueue.Job.safeRun(Job.java:229) 
                                                                              at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354) 
                                                                              at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36) 
                                                                              at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389) 
                                                                              at java.lang.Thread.run(Thread.java:818) 
`

**Comment by PatrickStoklasa** [comment]
Just as an update: We have experienced the issues as well and are currently fixem them as well as some other things and i expect a release to happen tomorrow.

**Comment by donpwilson** [comment]
Thanks. I can wait!

**Comment by donpwilson** [comment]
I have tested OneDrive, Dropbox and Box on 2.16.4.

In my first test I used getChildren(0 and all 3 interfaces worked correctly on my data.

I then switched to getChildrenPage() using the code I listed above. Box worked correctly, which validates my coding. 

However, both OneDrive and Dropbox crashed. I added a further tracing logcat message to help with debugging.

For OneDrive this showed that the search of the root folder found a folder called "Email attachments". I don't use this folder. As you can see above, my code loops to search sub-folders. It crashed when it called getChildrenPage() on this folder.This was the same for both attempts made.

```javaScript04-03 15:40:37.281 26293-27255/com.dpw.photoplayer I/Job_LoadCloudRailData: job-queue-worker-6ac shouldReRunOnThrowable() run count 1 drive OneDrive
                                                                            com.cloudrail.si.exceptions.NotFoundException: Item does not exist
                                                                                at com.cloudrail.si.services.OneDrive.getChildrenPage(OneDrive.java:1285)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:265)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:277)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111)
                                                                                at com.birbit.android.jobqueue.Job.safeRun(Job.java:229)
                                                                                at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354)
                                                                                at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389)
                                                                                at java.lang.Thread.run(Thread.java:818)
```
For DropBox , the crash occurred the first time a sub-folder was read. I know that this folder contains photos. My app makes two attempts and on the second attempt it crashed at the first call to getChildrenPage() on the root folder. To me this suggests that something in your code is not being properly re-initialized. I don't quite understand what you mean by "reproduce the relevant part of the file tree as well if you have spotted the issue."

```javaScript04-03 15:40:40.738 26293-27255/com.dpw.photoplayer I/Job_LoadCloudRailData: job-queue-worker-6ac shouldReRunOnThrowable() run count 1 drive Dropbox
                                                                            java.lang.RuntimeException: ServiceCode Error in function getChildrenPage at 25
                                                                                at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:111)
                                                                                at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45)
                                                                                at com.cloudrail.si.services.Dropbox.getChildrenPage(Dropbox.java:1029)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:265)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:277)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111)
                                                                                at com.birbit.android.jobqueue.Job.safeRun(Job.java:229)
                                                                                at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354)
                                                                                at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389)
                                                                                at java.lang.Thread.run(Thread.java:818)
                                                                             Caused by: java.lang.ArrayIndexOutOfBoundsException: length=12; index=-11
                                                                                at java.util.ArrayList.get(ArrayList.java:310)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:200)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:240)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:240)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getVariable(Sandbox.java:282)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getVariable(Sandbox.java:259)
                                                                                at com.cloudrail.si.servicecode.commands.Get.execute(Get.java:44)
                                                                                at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:87)
                                                                                at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45) 
                                                                                at com.cloudrail.si.services.Dropbox.getChildrenPage(Dropbox.java:1029) 
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:265) 
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:277) 
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111) 
                                                                                at com.birbit.android.jobqueue.Job.safeRun(Job.java:229) 
                                                                                at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354) 
                                                                                at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389) 
                                                                                at java.lang.Thread.run(Thread.java:818) 
04-03 15:40:40.742 26293-27121/com.dpw.photoplayer I/Job_LoadCloudRailData: job-queue-worker-740 onRun() source Dropbox further sources 1
04-03 15:40:40.744 26293-27121/com.dpw.photoplayer I/My_Photo_Records: job-queue-worker-740 740 searchFolders() folder / offset 0 limit 3
04-03 15:40:41.457 26293-27121/com.dpw.photoplayer I/Job_LoadCloudRailData: job-queue-worker-740 shouldReRunOnThrowable() run count 2 drive Dropbox
                                                                            java.lang.RuntimeException: ServiceCode Error in function getChildrenPage at 25
                                                                                at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:111)
                                                                                at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45)
                                                                                at com.cloudrail.si.services.Dropbox.getChildrenPage(Dropbox.java:1029)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:265)
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111)
                                                                                at com.birbit.android.jobqueue.Job.safeRun(Job.java:229)
                                                                                at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354)
                                                                                at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36)
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389)
                                                                                at java.lang.Thread.run(Thread.java:818)
                                                                             Caused by: java.lang.ArrayIndexOutOfBoundsException: length=12; index=-11
                                                                                at java.util.ArrayList.get(ArrayList.java:310)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:200)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:240)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getEntry(Sandbox.java:240)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getVariable(Sandbox.java:282)
                                                                                at com.cloudrail.si.servicecode.Sandbox.getVariable(Sandbox.java:259)
                                                                                at com.cloudrail.si.servicecode.commands.Get.execute(Get.java:44)
                                                                                at com.cloudrail.si.servicecode.Interpreter.run(Interpreter.java:87)
                                                                                at com.cloudrail.si.servicecode.Interpreter.callFunction(Interpreter.java:45) 
                                                                                at com.cloudrail.si.services.Dropbox.getChildrenPage(Dropbox.java:1029) 
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.searchFolders(Job_LoadCloudRailData.java:265) 
                                                                                at com.dpw.photoplayer.jobs.Job_LoadCloudRailData.onRun(Job_LoadCloudRailData.java:111) 
                                                                                at com.birbit.android.jobqueue.Job.safeRun(Job.java:229) 
                                                                                at com.birbit.android.jobqueue.JobHolder.safeRun(JobHolder.java:132) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.handleRunJob(ConsumerManager.java:407) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.access$000(ConsumerManager.java:326) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer$2.handleMessage(ConsumerManager.java:354) 
                                                                                at com.birbit.android.jobqueue.messaging.SafeMessageQueue.consume(SafeMessageQueue.java:36) 
                                                                                at com.birbit.android.jobqueue.ConsumerManager$Consumer.run(ConsumerManager.java:389) 
                                                                                at java.lang.Thread.run(Thread.java:818) 
``` 
Please let me know if I can help some more.

**Comment by PatrickStoklasa** [comment]
Now I was able to reproduce it. It is fixed and will be part of the next release.

**Comment by donpwilson** [comment]
Great! Glad I could be of help.

**Comment by PatrickStoklasa** [comment]
The fix for the above issue has now been released with version 2.16.5. Let me know if it works for you.

**Comment by donpwilson** [comment]
I confirm that 2.16.5 fixes this issue for my app,
